### PR TITLE
GCP authenticator - Conjur integration tests #1737

### PR DESCRIPTION
**Pull request by eranha**

Implementation of authn-gce integration tests of **input** validation and **configuration** tests.
Add `--authn-gce` flag support to the following scripts:

- ci/test
- dev/cli

The addition to the scripts sets the GCE authenticator test environment variables.

Connected to #[relevant GitHub issues](https://app.zenhub.com/workspaces/palmtree-5d99d900491c060001c85cba/issues/cyberark/conjur/1737)

### CODE REVIEW

File: cucumber/authenticators_gce/features/authn_gce.feature [hunk 1]
```diff
@@ -33,12 +33,41 @@ Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
     And I set "authn-gce/service-account-email" annotation to host "test-app"
     And I set "authn-gce/project-id" annotation to host "test-app"
     And I set "authn-gce/instance-name" annotation to host "test-app"
-    And I obtain a GCE identity token in "full" format with audience claim value: "conjur/cucumber/host/test-app"
+    And I obtain a GCE identity token in full format with audience claim value: "conjur/cucumber/host/test-app"
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce.feature:36]: why did you remove the quotes around `full`?

File: cucumber/authenticators_gce/features/authn_gce.feature [hunk 2]
```diff
@@ -33,12 +33,41 @@ Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
...
     And I save my place in the audit log file
-    When I authenticate with authn-gce using Google identity token
+    When I authenticate with authn-gce using token and account "cucumber"
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce.feature:38]: i would use `cucumber` as the default account so you don't need to send it each time. the regex `and account ""` can be optional

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 1 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/authn_gce.feature:36]: because it possible and much cleaner.

**Review by nessiLahav — APPROVED** [review]
👍

File: cucumber/authenticators_gce/features/authn_gce.feature [hunk 3]
```diff
@@ -33,12 +33,41 @@ Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
     And I set "authn-gce/service-account-email" annotation to host "test-app"
     And I set "authn-gce/project-id" annotation to host "test-app"
     And I set "authn-gce/instance-name" annotation to host "test-app"
-    And I obtain a GCE identity token in "full" format with audience claim value: "conjur/cucumber/host/test-app"
+    And I obtain a GCE identity token in full format with audience claim value: "conjur/cucumber/host/test-app"
     And I save my place in the audit log file
```
Comment (eladkug) [inline cucumber/authenticators_gce/features/authn_gce.feature:37]: we need to remove audit, since we dont have audit in conjur and it may confuse.
` And I save my place in the log file`

File: cucumber/authenticators_gce/features/authn_gce.feature [hunk 4]
```diff
@@ -33,12 +33,41 @@ Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
...
-    When I authenticate with authn-gce using Google identity token
+    When I authenticate with authn-gce using token and account "cucumber"
     Then host "test-app" has been authorized by Conjur
     And I can GET "/secrets/cucumber/variable/test-variable" with authorized user
     And The following appears in the audit log after my savepoint:
```
Comment (eladkug) [inline cucumber/authenticators_gce/features/authn_gce.feature:41]: same issue - remove audit

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (orenbm) [reply cucumber/authenticators_gce/features/authn_gce.feature:37]: we don't have audit in conjur? doesn't it get written to the log?

if so, we need to remove this from all the tests that verify audit, not just here

File: cucumber/authenticators_gce/features/authn_gce_bad_configuration.feature [hunk 5]
```diff
@@ -0,0 +1,57 @@
+Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce_bad_configuration.feature:1]: please update feature title

File: cucumber/authenticators_gce/features/authn_gce_hosts.feature [hunk 6]
```diff
@@ -0,0 +1,32 @@
+Feature: GCP Authenticator - Different Hosts can authenticate with GCE authenticator
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce_hosts.feature:1]: not all the hosts in the test can authenticate. please update.

File: cucumber/authenticators_gce/features/authn_gce_hosts.feature [hunk 7]
```diff
@@ -0,0 +1,32 @@
...
+      - !webservice
+
+      - !group apps
+
+      - !permit
+        role: !group apps
+        privilege: [ read, authenticate ]
+        resource: !webservice
+    """
+    And I am the super-user
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce_hosts.feature:21]: why do you need to be a super user? is this for future tests? (e.g create a host with different annotations)

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb [hunk 8]
```diff
@@ -14,19 +15,44 @@
     raise "Incorrect annotation name: '#{annotation_name}', expected: service-account-id|service-account-email|project-id|instance-name"
   end
 
-  set_annotation_to_resource("#{annotation_name}", annotation_value)
+  set_annotation_to_resource("authn-gce/#{annotation_name}", annotation_value)
 end
 
-Given(/^I obtain a GCE identity token in "([^"]*)" format with audience claim value: "([^"]*)"$/) do | audience, format |
+# Sets all GCE annotations
+Given(/^I set GCE annotations to host "([^"]*)"$/) do | hostname |
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:22]: why aren't you using this in the basic vanilla test?

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb [hunk 9]
```diff
@@ -14,19 +15,44 @@
...
+Given(/^I obtain a GCE identity token in (full|standard) format with audience claim value: "([^"]*)"$/) do | format, audience |
   gce_identity_access_token(
     audience: audience,
     token_format: format
   )
 end
 
-Given(/I authenticate with authn-gce using Google identity token/) do
+# Authenticates with Conjur GCE authenticator
+Given(/I authenticate with authn-gce using (no |an empty |invalid )?token and (non )?existing account/) do | token_state, non_existing_account |
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:40]: i think that `an` is redundant. but if you do use it, then use it also for `invalid`.

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb [hunk 10]
```diff
@@ -14,19 +15,44 @@
...
+  account = non_existing_account ? 'non-existing' : 'cucumber'
+
+  token = case token_state
+          when "no "
+            nil
+          when "empty "
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:46]: this should be `an empty`, no?

File: cucumber/authenticators_gce/features/support/authn_gce_helper.rb [hunk 11]
```diff
@@ -22,6 +22,10 @@ module AuthnGceHelper
   def gce_identity_access_token(audience:, token_format: 'standard')
     audience = audience.gsub("/", "%2F")
 
+    if(!File.exist?(private_key_path))
+      raise  "GCE private key credentials file #{private_key_path} not found."
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/support/authn_gce_helper.rb:26]: you have 2 spaces after `raise`

File: cucumber/authenticators_gce/features/support/authn_gce_helper.rb [hunk 12]
```diff
@@ -22,6 +22,10 @@ module AuthnGceHelper
   def gce_identity_access_token(audience:, token_format: 'standard')
     audience = audience.gsub("/", "%2F")
 
+    if(!File.exist?(private_key_path))
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/support/authn_gce_helper.rb:25]: remove parantheses + use `unless` instead of `if !`

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb (hunk 8 above)
Comment (eladkug) [inline cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:22]: Do we want it will verify all annotations ? i prefer it will be one by one, otherwise its not clear by reading the feature file only.

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/authn_gce.feature:37]: here it's ` And I save my place in the audit log file`
on you comment it's ` And I save my place in the log file` without `audit`

**Review by eladkug — CHANGES_REQUESTED** [review]
Added some changes.. Some tests are missing here, by i understood there is a different tickets for this. 
Examples :
Verify specific annotations (here we verify all 4)
Performance
Status...

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (orenbm) [reply cucumber/authenticators_gce/features/authn_gce.feature:37]: @eladkug let's do this in another PR as it's out of scope of GCE. we have tests that verify the audit log but you say that there isn't one. 

i think that what happens here is that the tests assign a socket for audit and verifies that the audit is written there. this seems to be a good thing but we can challenge that. anyway, it should be outside of this PR imo

File: cucumber/authenticators_gce/features/authn_gce_hosts.feature (hunk 7 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/authn_gce_hosts.feature:21]: nope.

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb (hunk 8 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:22]: Wanted to show the full set.

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb (hunk 9 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:40]: I agree.

File: cucumber/authenticators_gce/features/support/authn_gce_helper.rb (hunk 11 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/support/authn_gce_helper.rb:26]: I like you

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb (hunk 8 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:22]: There's also a method to set explicit.

Comment (eladkug) [reply cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb:22]: than maybe change the name to obtain all CGE annotations to host

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (orenbm) [reply cucumber/authenticators_gce/features/authn_gce.feature:37]: is this the only blocker on this PR? @eladkug can we address this in another issue as it's not GCE realted?

File: cucumber/authenticators_gce/features/authn_gce.feature [hunk 13]
```diff
@@ -33,12 +33,41 @@ Feature: GCE Authenticator - Hosts can authenticate with GCE authenticator
...
+
+  Scenario: Missing GCE access token is a bad request
+    Given I save my place in the log file
+    When I authenticate with authn-gce using no token and existing account
+    Then it is a bad request
+    And The following appears in the log after my savepoint:
+    """
+    Errors::Authentication::RequestBody::MissingRequestParam
+    """
+
```
Comment (orenbm) [inline cucumber/authenticators_gce/features/authn_gce.feature:54]: 2 empty lines instead of 1

File: cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb (hunk 8 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/step_definitions/authn_gce_steps.rb]: You probably meant  `set all...` NP.

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 13 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/authn_gce.feature:51]: Yea mama.

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (eladkug) [reply cucumber/authenticators_gce/features/authn_gce.feature:34]: I agree it can be on different PR. @eranha please open a ticket for this

**Review by orenbm — APPROVED** [review]

**Review by orenbm — APPROVED** [review]

File: cucumber/authenticators_gce/features/authn_gce.feature (hunk 3 above)
Comment (eranha) [reply cucumber/authenticators_gce/features/authn_gce.feature]: The word `audit` is ignored in the step. I will create an issue to remove the word audit.
`# NOTE: The source code order of this step def and the one above matters.
And(/^The following appears in the (audit )?log after my savepoint:$/) do |_, message|
  expect(num_matches_since_savepoint(message)).to be >= 1
end`
